velox/functions/lib/aggregates/MinMaxByAggregatesBase.h

```diff
@@ -31,631 +34,297 @@ constexpr bool isNumeric() {
-    auto baseRowVector =
-        dynamic_cast<const RowVector*>(decodedIntermediateResult_.base());
-
-    decodedValue_.decode(*baseRowVector->childAt(0), rows);
-    decodedComparison_.decode(*baseRowVector->childAt(1), rows);
-
-    if (decodedIntermediateResult_.isConstantMapping()) {
-      if (decodedIntermediateResult_.isNullAt(0)) {
-        return;
+    void destroy(HashStringAllocator* allocator) {
+      if constexpr (!isNumeric<V>()) {
+        if (currValue_.has_value()) {
+          currValue_->destroy(allocator);
+        }
       }
-      const auto decodedIndex = decodedIntermediateResult_.index(0);
-      updateValues(
-          group,
-          decodedValue_,
-          decodedComparison_,
-          decodedIndex,
-          decodedValue_.isNullAt(decodedIndex),
-          mayUpdate);
-    } else if (decodedIntermediateResult_.mayHaveNulls()) {
-      rows.applyToSelected([&](vector_size_t i) {
-        if (decodedIntermediateResult_.isNullAt(i)) {
-          return;
+
+      if constexpr (!isNumeric<C>()) {
+        if (currComparison_.has_value()) {
+          currComparison_->destroy(allocator);
         }
-        const auto decodedIndex = decodedIntermediateResult_.index(i);
-        updateValues(
-            group,
-            decodedValue_,
-            decodedComparison_,
-            decodedIndex,
-            decodedValue_.isNullAt(decodedIndex),
-            mayUpdate);
-      });
-    } else {
-      rows.applyToSelected([&](vector_size_t i) {
-        const auto decodedIndex = decodedIntermediateResult_.index(i);
-        updateValues(
-            group,
-            decodedValue_,
-            decodedComparison_,
-            decodedIndex,
-            decodedValue_.isNullAt(decodedIndex),
-            mayUpdate);
-      });
+      }
     }
-  }
-
-  void initializeNewGroupsInternal(
-      char** groups,
-      folly::Range<const vector_size_t*> indices) override {
-    exec::Aggregate::setAllNulls(groups, indices);
-    for (const vector_size_t i : indices) {
-      auto group = groups[i];
-      valueIsNull(group) = true;
 
-      if constexpr (!isNumeric<T>()) {
-        new (group + offset_) SingleValueAccumulator();
-      } else {
-        *value(group) = ValueAccumulatorType();
+   private:
+    bool needUpdate(const exec::GenericView& newComparison) {
+      if (!currComparison_.has_value()) {
+        return true;
       }
 
-      if constexpr (isNumeric<U>()) {
-        *comparisonValue(group) = ComparisonAccumulatorType();
+      if constexpr (isNumeric<C>()) {
+        return Comparator<isMaxFunc, C, ComparisonAccumulatorType>::compare(
+            currComparison_.value(), newComparison.castTo<C>());
       } else {
-        new (group + offset_ + sizeof(ValueAccumulatorType))
-            SingleValueAccumulator();
+        return Comparator<isMaxFunc, C, ComparisonAccumulatorType>::compare(
+            currComparison_.value(),
+            newComparison.decoded(),
```

Comment:
Curious, If you can call .value() on currComparison, why not get reference to the value from newComparison without having to pass a reference to decoded vector etc ?

Reply:
It's for the following intuition:
1. `currComparison_` might be null so I have to check whether it has value on line 184
2. `currComparison_` internally has type of `SingleValueAccumulator` for complex types, and `SingleValueAccumulator` has helper function `compare` which can take reference of decoded vector for comparing.